prov/rxm/src/rxm_conn.c

```diff
@@ -360,9 +385,10 @@ static void rxm_conn_handle_eq_err(struct rxm_ep *rxm_ep, ssize_t rd)
 	}
 	OFI_EQ_READERR(&rxm_prov, FI_LOG_FABRIC, rxm_ep->msg_eq, rd, err_entry);
 	if (err_entry.err == ECONNREFUSED) {
-		FI_DBG(&rxm_prov, FI_LOG_FABRIC, "Connection refused\n");
-		ofi_cmap_process_reject(rxm_ep->util_ep.cmap,
-					err_entry.fid->context);
+		struct rxm_conn *rxm_conn =
+			container_of(err_entry.fid->context,
+				     struct rxm_conn, handle);
```

Comment:
Here rxm_conn may have a handle with values updated with a valid connection info, in case of simultaneous connections. It may not have info about refused msg endpoint. Because of that you may end up closing the wrong msg ep while processing UTIL_CMAP_CMD_CONN_CLOSE in cmd queue. This is the reason why you are getting FI_SHUTDOWN from tcp provider in certain cases.

Reply:
Yes, right. A handle can contain a new connection info. We use `close_data` field from `rxm_conn` that carries MSG EP that should be closed

Reply:
At the time of receiving FI_REJECT, `rxm_conn->msg_ep` would already be overwritten with new  `msg_ep` that was created by accepting connection request from remote. This `msg_ep` should not be closed. En-queuing `UTIL_CMAP_CMD_CONN_CLOSE` command should be done as part of `rxm_conn_close`.

Reply:
Sorry. On a second glance, I think we're okay here.